merlin/models/tf/core.py

```diff
@@ -2504,57 +2540,106 @@ def query_embeddings(
+        query_id_tag: str
+            Tag to use for the query id.
         batch_size: int
             Batch size to use for embedding extraction.
 
         Returns
         -------
         merlin.io.Dataset
         """
-        from merlin.models.tf.prediction.batch import QueryEmbeddings
-
-        get_user_emb = QueryEmbeddings(self, dim=dim, batch_size=batch_size)
+        from merlin.models.tf.utils.batch_utils import QueryEmbeddings
 
-        # Check if merlin-dataset is passed
-        if hasattr(dataset, "to_ddf"):
-            dataset = dataset.to_ddf()
+        get_user_emb = QueryEmbeddings(self, batch_size=batch_size)
 
+        dataset = self._ensure_unique(dataset, query_tag, query_id_tag)
         embeddings = dataset.map_partitions(get_user_emb)
 
         return merlin.io.Dataset(embeddings)
 
     def item_embeddings(
-        self, dataset: merlin.io.Dataset, dim: int, batch_size=None, **kwargs
+        self,
+        dataset: merlin.io.Dataset,
+        item_tag=Tags.ITEM,
+        item_id_tag=Tags.ITEM_ID,
+        batch_size=None,
     ) -> merlin.io.Dataset:
         """Export item embeddings from the model.
 
         Parameters
         ----------
         dataset: merlin.io.Dataset
             Dataset to export embeddings from.
-        dim: int
-            Dimensionality of the embeddings.
+        item_tag: str
+            Tag to use for the item.
+        item_id_tag: str
+            Tag to use for the item id.
         batch_size: int
             Batch size to use for embedding extraction.
 
         Returns
         -------
         merlin.io.Dataset
         """
-        from merlin.models.tf.prediction.batch import ItemEmbeddings
-
-        get_item_emb = ItemEmbeddings(self, dim=dim, batch_size=batch_size)
+        from merlin.models.tf.utils.batch_utils import ItemEmbeddings
 
-        # Check if merlin-dataset is passed
-        if hasattr(dataset, "to_ddf"):
-            dataset = dataset.to_ddf()
+        get_item_emb = ItemEmbeddings(self, batch_size=batch_size)
 
+        dataset = self._ensure_unique(dataset, item_tag, item_id_tag)
         embeddings = dataset.map_partitions(get_item_emb)
 
         return merlin.io.Dataset(embeddings)
 
+    def load_topk_evaluation(self, data: merlin.io.Dataset, k: int, **kwargs):
+        """Update the model with a top-k evaluation block.
+
+        Parameters
+        ----------
+        data: merlin.io.Dataset
+            Dataset of negatives to use for evaluation.
+        k: int
+            Number of negatives to retrieve.
+
+        Returns
+        -------
+        SequentialBlock
+        """
+        import merlin.models.tf as ml
+
+        topk_index = ml.TopKIndexBlock.from_block(
+            self.retrieval_block.item_block(), data=data, k=k - 1, context=self.context, **kwargs
+        )
+        # set cache_query to True in the ItemRetrievalScorer
+        self.loss_block.retrieval_scorer.cache_query = True
```

Comment:
Would it make sense to have a method in `ItemRetrievalTask` (`loss_block`) to set `cache_query` and `pre_metrics`, rather than exposing those properties for other classes?

Reply:
The `loss_block` is a property of the general Model class, and I added the `retrieval_scorer` to the `ItemRetrievalTask`